### DISCUSSION TITLE
Customizing the titlebar - can I modify the decorations?

### DISCUSSION THREAD

**Discussion by timleg002** [tauri-apps/tauri#3093] [Q&A] [answered]
I am currently making a file manager app and I want to customize the titlebar, atleast making it transparent, but I only found out how to disable the decorations and implement the titlebar by myself. However when the decorations are turned off, there is no shadow coming from the app (for user to know it is there), I can't seem to implement a `border-radius` (only tried doing `style="border-radius: 12` on `body`) and didn't try to implement a border yet, but the shadow one seems impossible. I'm up for implementing this more direclty in tauri, just couldn't seem to find anything related to decorations that wasn't a loop to itself.

**Comment by amrbashir** [comment, marked as answer]
Disable decorations and implement your custom titlebar. An option for shadows will be added in tauri v2 for macOS and Windows. Rounded corners for now need to be done in css and you need to make your window transparent (on macOS it might disable shadows). In tauri v2 we will add an option for rounded corners (at least for macOS) but no guarantees.

**Reply by ple1n** [reply]
No shadow option on linux ? Wish we could do that.

**Reply by amrbashir** [reply]
Here is a [plugin ](https://github.com/tauri-apps/window-shadows) to add native shadows, it only works with macOS and Windows. On Linux, shadows are controlled by the compositor on the user machine, they need to explicitly set it for your window. It is a limitation of Linux and we can't do anything about it.

**Reply by ple1n** [reply]
Maybe we should file issues for compositors to change the default behaviour, directly. I'm not sure how, though.

**Reply by mrcoon** [reply]
Disable decorations makes it impossible to drag and drop windows. Is there any way to solve this problem?

**Reply by JonasKruckenberg** [reply]
You have to add the `data-tauri-drag-region` attribute to any html tags you wish to be "drag regions" for your app

**Reply by mrcoon** [reply]
> You have to add the `data-tauri-drag-region` attribute to any html tags you wish to be "drag regions" for your app

Great, thanks for the answer, that solved my problem. 
Then the function keys in the Title Bar (close/reduce/full screen) need to be implemented manually? Wouldn't it be more convenient if the Title Bar supports transparency?

**Reply by JonasKruckenberg** [reply]
Yeah it supports transparency, though not in core (yet) the code is here: https://github.com/tauri-apps/tauri/issues/2663#issuecomment-1088906114

And the PR to merge that is here: https://github.com/tauri-apps/tauri/pull/3965

**Reply by alectrocute** [reply]
@JonasKruckenberg Do we have `data-tauri-drag-region` in the docs? I had to hunt to find this answer.

**Comment by timleg002** [comment]
Great, thanks! How would I create the rounded corners via CSS, as you
mentioned?

**Reply by amrbashir** [reply]
Make your window transparent and 
and use this css:
```css
html, body {
  border-radius: 10px; /* whatever you want */
  background-color: #fff; /* your app bg color */
}
```

**Reply by timleg002** [reply]
Unfortunately this doesn't work, I still have sharp corners

**Reply by timleg002** [reply]
I also noticed that you wrote a plugin that adds the vibrancy, by that you mean the shadows?

**Reply by amrbashir** [reply]
That must be the content inside the body.  Try this instead
```css
html, body {
  background-color: #fff;
  border-radius: 10px;
  overflow: hidden;
}

.element-inside-your-body {
  overflow: auto;
}
```
This is general css stuff and not tauri related btw.

**Reply by timleg002** [reply]
Still doesn't work, for context I'm using react and the `.element-inside-your-body` would be `#root`. The corners are still sharp

**Reply by amrbashir** [reply]
As I said, this is a general css stuff. I can't possibly debug your code until I see it. And don't forget to make your window transparent.

As for the vibrancy plugin it is for blur effects not shadows.

**Reply by timleg002** [reply]
Thanks for the help! I'll probably lay it off for now

**Reply by M4rteaga** [reply]
In order for this to work you need to make your window transparent and set up macOSPrivateApi true (WARNING: Using private APIs on macOS prevents your application from being accepted to the App Store [docs](https://tauri.app/v1/api/config#windowconfig)). After that you should add the styles to index.html and it should work

``tauri.conf.json``
```json 
{
  "windows": [{
            "label": "main",
            "fullscreen": false,
            "height": 540,
            "resizable": false,
            "title": "todo-app",
            "width": 320,
            "visible": false,
            "alwaysOnTop": true,
            "decorations": false,
            "transparent": true
        }],
  "macOSPrivateApi": true
}
```

```/index.html inside the head tag```
```html
 <style>
    html,
    body {
      border-radius: 10px;
      overflow: hidden;
    }

    .element-inside-your-body {
      overflow: auto;
    }
  </style>
```

It ended up looking like this
![CleanShot 2022-11-28 at 23 01 59](https://user-images.githubusercontent.com/47699470/204435797-cd7bab72-8ae4-4340-ac99-5b1ebf96adf1.png)

**Reply by SaharshSamir** [reply]
When i try to do that, i get these weird behaviours around the corners like: 
![image](https://user-images.githubusercontent.com/45710871/228677261-a84b9c4d-2162-4cff-9962-930a6df30291.png)

**Reply by jzillmann** [reply]
Sure you configured `"macOSPrivateApi": true` ?

**Reply by lonffreytu** [reply]
> In order for this to work you need to make your window transparent and set up macOSPrivateApi true (WARNING: Using private APIs on macOS prevents your application from being accepted to the App Store [docs](https://tauri.app/v1/api/config#windowconfig)). After that you should add the styles to index.html and it should work
> 
> `tauri.conf.json`
> 
> ```json
> {
>   "windows": [{
>             "label": "main",
>             "fullscreen": false,
>             "height": 540,
>             "resizable": false,
>             "title": "todo-app",
>             "width": 320,
>             "visible": false,
>             "alwaysOnTop": true,
>             "decorations": false,
>             "transparent": true
>         }],
>   "macOSPrivateApi": true
> }
> ```
> 
> `/index.html inside the head tag`
> 
> ```
>  <style>
>     html,
>     body {
>       border-radius: 10px;
>       overflow: hidden;
>     }
> 
>     .element-inside-your-body {
>       overflow: auto;
>     }
>   </style>
> ```
> 
> It ended up looking like this ![CleanShot 2022-11-28 at 23 01 59](https://user-images.githubusercontent.com/47699470/204435797-cd7bab72-8ae4-4340-ac99-5b1ebf96adf1.png)

It seems that it doesn't work on Windows 10.
